private/model/api/api.go

```diff
@@ -598,7 +607,14 @@ func newClient(cfg aws.Config, handlers request.Handlers, partitionID, endpoint,
 	svc.Handlers.Build.PushBackNamed({{ .ProtocolPackage }}.BuildHandler)
 	svc.Handlers.Unmarshal.PushBackNamed({{ .ProtocolPackage }}.UnmarshalHandler)
 	svc.Handlers.UnmarshalMeta.PushBackNamed({{ .ProtocolPackage }}.UnmarshalMetaHandler)
-	svc.Handlers.UnmarshalError.PushBackNamed({{ .ProtocolPackage }}.UnmarshalErrorHandler)
+	{{- if and $.WithGeneratedTypedErrors (gt (len $.ShapeListErrors) 0) }}
+		{{- $_ := $.AddSDKImport "private/protocol" }}
+		svc.Handlers.UnmarshalError.PushBackNamed(
+			protocol.NewUnmarshalErrorHandler({{ .ProtocolPackage }}.NewUnmarshalTypedError(exceptionFromCode)).NamedHandler(),
```

Comment:
This constructor adds syntactical sugar but isn't needed, as `UnmarshalErrorHandler` is exported in protocol package. 

We should directly call  `UnmarshalErrorHandler{}`, this also makes it consistent with marshaler design.

Reply:
I think constructors are generally preferred when a value requires input parameters to be initialized. In this case both types use their constructor to take a input value and assign it to a unexported member. If the constructors were not present the member must be made exported, and becomes a part of the public interface.

Reply:
With that in mind the v2 SDK's generated marshaler design could be updated to take the `input` value as a parameter to a constructor. Since the value would be required.  But that case is slightly different since all marshaler types are unexported and there is no concern about public interface being exposed.